File: frontend/js/views/rule_of_78.js
// Rule of 78s — precomputed-interest rebate, via /calc/rule-of-78.
import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
export async function renderRuleOf78(mount, _s) {
    const tok = currentViewToken(); if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.r78.h1.title">// RULE OF 78s</span></h1>
        <p class="muted small" data-i18n="view.r78.hint.intro">How a precomputed-interest loan allocates its finance charge and the rebate on early payoff. Interest is front-loaded by the sum-of-the-digits method (SOD = n(n+1)/2), so paying off early refunds less unearned interest than simple pro-rata — the penalty this quantifies. Not financial advice.</p>
        <div class="lpv-split"><div class="chart-panel"><h2 data-i18n="view.r78.h2.inputs">Loan</h2>
        <form id="r78-form" class="inline-form">
            <label><span data-i18n="view.r78.label.charge">Total finance charge ($)</span><input type="number" step="10" min="0" name="total_finance_charge_usd" value="1200" required></label>
            <label><span data-i18n="view.r78.label.term">Original term (months)</span><input type="number" step="1" min="1" name="original_term_months" value="36" required></label>
            <label><span data-i18n="view.r78.label.made">Payments made</span><input type="number" step="1" min="0" name="payments_made" value="12"></label>
        </form></div><div id="r78-result" class="chart-panel lpv-preview"></div></div>`;
    applyUiI18n(mount);
    const form = mount.querySelector('#r78-form'); const n = (k) => Number(form.querySelector(`[name="${k}"]`).value) || 0;
    const gen = async () => {
        const body = { total_finance_charge_usd: n('total_finance_charge_usd'), original_term_months: n('original_term_months'), payments_made: n('payments_made') };
        try { const d = await api.calcRuleOf78(body); if (!viewIsCurrent(tok)) return; res(mount, d); }
        catch (e) { showToast(e.message || t('view.r78.toast.error'), { level: 'error' }); }
    };
    const live = debounce(gen, 250); form.addEventListener('input', () => live()); form.addEventListener('submit', (e) => { e.preventDefault(); gen(); }); gen();
}
function res(mount, d) {
    const el = mount.querySelector('#r78-result');
    if (!d.valid) { el.innerHTML = `<p class="muted" data-i18n="view.r78.invalid">Payments made cannot exceed the term.</p>`; applyUiI18n(el); return; }
    el.innerHTML = `<div class="lpv-bar"><div class="cards">
        <div class="card pos"><div class="label" data-i18n="view.r78.card.rebate">Rebate (unearned)</div><div class="value">${money(d.rebate_usd)}</div></div>
        <div class="card"><div class="label" data-i18n="view.r78.card.earned">Earned interest</div><div class="value">${money(d.earned_interest_usd)}</div></div>
        <div class="card"><div class="label" data-i18n="view.r78.card.straight">Pro-rata earned</div><div class="value">${money(d.straight_line_earned_usd)}</div></div>
        <div class="card neg"><div class="label" data-i18n="view.r78.card.penalty">Early-payoff penalty</div><div class="value">${money(d.early_payoff_penalty_usd)}</div></div>
        <div class="card"><div class="label" data-i18n="view.r78.card.sod">Sum of digits</div><div class="value">${d.sum_of_digits}</div></div>
    </div></div>`;
    applyUiI18n(el);
}

File: crates/traderview-core/src/rule_of_78.rs
//! Rule of 78s (sum-of-the-digits interest rebate) — how a precomputed-interest
//! loan allocates its finance charge and what rebate a borrower gets on early
//! payoff. The total finance charge is spread over the term weighted by the
//! remaining months: month 1 carries n/SOD of the interest, the last month 1/SOD,
//! where SOD = n(n+1)/2. On payoff after m payments the unearned interest
//! (rebate) is `F × (n−m)(n−m+1) / (n(n+1))`. Because it front-loads interest, the
//! borrower earns a smaller rebate than simple pro-rata — the early-payoff
//! penalty this module quantifies. Not financial advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct RuleOf78Input {
    /// Total precomputed finance charge over the full term.
    pub total_finance_charge_usd: f64,
    /// Original loan term in months.
    pub original_term_months: u32,
    /// Payments already made at the point of payoff.
    #[serde(default)]
    pub payments_made: u32,
}

#[derive(Debug, Clone, Serialize, PartialEq, Default)]
pub struct RuleOf78Report {
    /// Sum of the digits, n(n+1)/2.
    pub sum_of_digits: u32,
    /// Unearned interest refunded on early payoff.
    pub rebate_usd: f64,
    /// Finance charge the lender keeps (total − rebate).
    pub earned_interest_usd: f64,
    /// Pro-rata (straight-line) earned interest for comparison.
    pub straight_line_earned_usd: f64,
    /// Earned − straight-line: the extra cost of the Rule-of-78 front-loading.
    pub early_payoff_penalty_usd: f64,
    pub valid: bool,
}

fn cents(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

pub fn generate(i: &RuleOf78Input) -> RuleOf78Report {
    let n = i.original_term_months;
    if n == 0 || i.payments_made > n {
        return RuleOf78Report::default();
    }
    let m = i.payments_made;
    let sod = n * (n + 1) / 2;
    let rem = n - m;
    let f = i.total_finance_charge_usd;
    // Unearned fraction = remaining-months' digit sum ÷ total digit sum.
    let rebate = f * (rem as f64 * (rem as f64 + 1.0)) / (n as f64 * (n as f64 + 1.0));
    let earned = f - rebate;
    let straight = f * m as f64 / n as f64;
    RuleOf78Report {
        sum_of_digits: sod,
        rebate_usd: cents(rebate),
        earned_interest_usd: cents(earned),
        straight_line_earned_usd: cents(straight),
        early_payoff_penalty_usd: cents(earned - straight),
        valid: true,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn base() -> RuleOf78Input {
        RuleOf78Input {
            total_finance_charge_usd: 1_200.0,
            original_term_months: 36,
            payments_made: 12,
        }
    }

    #[test]
    fn rebate_and_penalty() {
        let d = generate(&base());
        assert_eq!(d.sum_of_digits, 666);
        assert!(close(d.rebate_usd, 540.54));
        assert!(close(d.earned_interest_usd, 659.46));
        assert!(close(d.straight_line_earned_usd, 400.0));
        assert!(close(d.early_payoff_penalty_usd, 259.46));
    }

    #[test]
    fn full_term_no_rebate() {
        let d = generate(&RuleOf78Input { payments_made: 36, ..base() });
        assert!(close(d.rebate_usd, 0.0));
        assert!(close(d.earned_interest_usd, 1_200.0));
    }

    #[test]
    fn no_payments_full_rebate() {
        let d = generate(&RuleOf78Input { payments_made: 0, ..base() });
        assert!(close(d.rebate_usd, 1_200.0));
        assert!(close(d.earned_interest_usd, 0.0));
    }

    #[test]
    fn front_loaded_penalty_positive_early() {
        // Early payoff always earns the lender more than straight-line.
        let d = generate(&base());
        assert!(d.early_payoff_penalty_usd > 0.0);
    }

    #[test]
    fn invalid_inputs() {
        assert!(!generate(&RuleOf78Input { original_term_months: 0, ..base() }).valid);
        assert!(!generate(&RuleOf78Input { payments_made: 40, ..base() }).valid);
    }
}
